File: practice-java/fundamentals/src/Operadores.java
public class Operadores {
    public static void main(String[] args) {

        // ===== OPERADORES ARITMÉTICOS =====

        int numero1 = 20;
        int numero2 = 5;

        int suma = numero1 + numero2;
        int resta = numero1 - numero2;
        int multiplicacion = numero1 * numero2;
        int division = numero1 / numero2;
        int modulo = numero1 % numero2; // residuo de la división

        System.out.println("OPERADORES ARITMÉTICOS:");
        System.out.println(numero1 + " + " + numero2 + " = " + suma);
        System.out.println(numero1 + " - " + numero2 + " = " + resta);
        System.out.println(numero1 + " * " + numero2 + " = " + multiplicacion);
        System.out.println(numero1 + " / " + numero2 + " = " + division);
        System.out.println(numero1 + " % " + numero2 + " = " + modulo);

        // ===== OPERADORES RELACIONALES (COMPARACIÓN) =====

        System.out.println("\nOPERADORES RELACIONALES:");
        System.out.println(numero1 + " == " + numero2 + " : " + (numero1 == numero2)); // igual a
        System.out.println(numero1 + " != " + numero2 + " : " + (numero1 != numero2)); // distinto a
        System.out.println(numero1 + " > " + numero2 + " : " + (numero1 > numero2));   // mayor que
        System.out.println(numero1 + " < " + numero2 + " : " + (numero1 < numero2));   // menor que
        System.out.println(numero1 + " >= " + numero2 + " : " + (numero1 >= numero2)); // mayor o igual
        System.out.println(numero1 + " <= " + numero2 + " : " + (numero1 <= numero2)); // menor o igual

        // ===== OPERADORES LÓGICOS =====

        boolean condicion1 = true;
        boolean condicion2 = false;

        System.out.println("\nOPERADORES LÓGICOS:");
        System.out.println("true && false (AND): " + (condicion1 && condicion2));
        System.out.println("true || false (OR): " + (condicion1 || condicion2));
        System.out.println("!true (NOT): " + (!condicion1));

        // AND (&&): ambas condiciones deben ser verdaderas
        boolean loginExitoso = (numero1 > numero2) && (numero2 > 0); // true && true = true
        System.out.println("¿Login exitoso? " + loginExitoso);

        // OR (||): al menos una condición debe ser verdadera
        boolean esFeriadoOFindeSemana = false || true; // false || true = true
        System.out.println("¿Es feriado o fin de semana? " + esFeriadoOFindeSemana);

        // ===== OPERADORES DE ASIGNACIÓN =====

        System.out.println("\nOPERADORES DE ASIGNACIÓN:");
        int valor = 10;
        System.out.println("valor inicial: " + valor);

        valor += 5; // equivalente a: valor = valor + 5
        System.out.println("después de += 5: " + valor);

        valor -= 3; // equivalente a: valor = valor - 3
        System.out.println("después de -= 3: " + valor);

        valor *= 2; // equivalente a: valor = valor * 2
        System.out.println("después de *= 2: " + valor);

        valor /= 4; // equivalente a: valor = valor / 4
        System.out.println("después de /= 4: " + valor);

        // ===== OPERADORES DE INCREMENTO Y DECREMENTO =====

        System.out.println("\nOPERADORES DE INCREMENTO Y DECREMENTO:");
        int contador = 5;
        System.out.println("contador inicial: " + contador);

        contador++; // incremento post-fijo (usa el valor actual, luego incrementa)
        System.out.println("después de contador++: " + contador);

        contador--; // decremento post-fijo (usa el valor actual, luego decrementa)
        System.out.println("después de contador--: " + contador);

        ++contador; // incremento pre-fijo (incrementa primero, luego usa el valor)
        System.out.println("después de ++contador: " + contador);

        --contador; // decremento pre-fijo (decrementa primero, luego usa el valor)
        System.out.println("después de --contador: " + contador);

        // ===== OPERADOR TERNARIO =====

        System.out.println("\nOPERADOR TERNARIO:");
        int edad = 18;
        String puedeVotar = (edad >= 18) ? "Sí, puede votar" : "No, aún es menor de edad";
        System.out.println("¿" + edad + " años puede votar? " + puedeVotar);

        int calificacion = 75;
        String resultado = (calificacion >= 60) ? "Aprobado" : "Reprobado";
        System.out.println("Calificación de " + calificacion + ": " + resultado);

        // ===== OPERADORES BITWISE (BIT A BIT) =====

        System.out.println("\nOPERADORES BITWISE:");
        int a = 5;    // binario: 0101
        int b = 3;    // binario: 0011

        System.out.println("a = " + a + " (binario: " + Integer.toBinaryString(a) + ")");
        System.out.println("b = " + b + " (binario: " + Integer.toBinaryString(b) + ")");

        int and = a & b;        // AND bit a bit
        int or = a | b;         // OR bit a bit
        int xor = a ^ b;        // XOR bit a bit
        int not = ~a;           // NOT bit a bit
        int leftShift = a << 1; // desplazamiento a la izquierda
        int rightShift = a >> 1; // desplazamiento a la derecha

        System.out.println(a + " & " + b + " = " + and + " (binario: " + Integer.toBinaryString(and) + ")");
        System.out.println(a + " | " + b + " = " + or + " (binario: " + Integer.toBinaryString(or) + ")");
        System.out.println(a + " ^ " + b + " = " + xor + " (binario: " + Integer.toBinaryString(xor) + ")");
        System.out.println("~" + a + " = " + not);
        System.out.println(a + " << 1 = " + leftShift);
        System.out.println(a + " >> 1 = " + rightShift);

    }
}

File: practice-cpp/fundamentals/src/operadores.cpp
#include <iostream>
#include <iomanip>

using namespace std;

int main() {

    // ===== OPERADORES ARITMÉTICOS =====

    cout << "\n===== OPERADORES ARITMÉTICOS =====\n";

    int numero1 = 20;
    int numero2 = 5;

    int suma = numero1 + numero2;
    int resta = numero1 - numero2;
    int multiplicacion = numero1 * numero2;
    int division = numero1 / numero2;
    int modulo = numero1 % numero2; // residuo de la división

    cout << numero1 << " + " << numero2 << " = " << suma << "\n";
    cout << numero1 << " - " << numero2 << " = " << resta << "\n";
    cout << numero1 << " * " << numero2 << " = " << multiplicacion << "\n";
    cout << numero1 << " / " << numero2 << " = " << division << "\n";
    cout << numero1 << " % " << numero2 << " = " << modulo << "\n";

    // ===== OPERADORES RELACIONALES =====

    cout << "\n===== OPERADORES RELACIONALES =====\n";

    cout << numero1 << " == " << numero2 << " : " << (numero1 == numero2 ? "true" : "false") << "\n";
    cout << numero1 << " != " << numero2 << " : " << (numero1 != numero2 ? "true" : "false") << "\n";
    cout << numero1 << " > " << numero2 << " : " << (numero1 > numero2 ? "true" : "false") << "\n";
    cout << numero1 << " < " << numero2 << " : " << (numero1 < numero2 ? "true" : "false") << "\n";
    cout << numero1 << " >= " << numero2 << " : " << (numero1 >= numero2 ? "true" : "false") << "\n";
    cout << numero1 << " <= " << numero2 << " : " << (numero1 <= numero2 ? "true" : "false") << "\n";

    // ===== OPERADORES LÓGICOS =====

    cout << "\n===== OPERADORES LÓGICOS =====\n";

    bool condicion1 = true;
    bool condicion2 = false;

    cout << "true && false (AND): " << (condicion1 && condicion2 ? "true" : "false") << "\n";
    cout << "true || false (OR): " << (condicion1 || condicion2 ? "true" : "false") << "\n";
    cout << "!true (NOT): " << (!condicion1 ? "true" : "false") << "\n";

    // AND: ambas condiciones deben ser verdaderas
    bool loginExitoso = (numero1 > numero2) && (numero2 > 0);
    cout << "¿Login exitoso? " << (loginExitoso ? "true" : "false") << "\n";

    // OR: al menos una condición debe ser verdadera
    bool esFeriadoOFindeSemana = false || true;
    cout << "¿Es feriado o fin de semana? " << (esFeriadoOFindeSemana ? "true" : "false") << "\n";

    // ===== OPERADORES DE ASIGNACIÓN =====

    cout << "\n===== OPERADORES DE ASIGNACIÓN =====\n";

    int valor = 10;
    cout << "valor inicial: " << valor << "\n";

    valor += 5; // equivalente a: valor = valor + 5
    cout << "después de += 5: " << valor << "\n";

    valor -= 3; // equivalente a: valor = valor - 3
    cout << "después de -= 3: " << valor << "\n";

    valor *= 2; // equivalente a: valor = valor * 2
    cout << "después de *= 2: " << valor << "\n";

    valor /= 4; // equivalente a: valor = valor / 4
    cout << "después de /= 4: " << valor << "\n";

    // ===== OPERADORES DE INCREMENTO Y DECREMENTO =====

    cout << "\n===== OPERADORES DE INCREMENTO Y DECREMENTO =====\n";

    int contador = 5;
    cout << "contador inicial: " << contador << "\n";

    contador++; // incremento post-fijo
    cout << "después de contador++: " << contador << "\n";

    contador--; // decremento post-fijo
    cout << "después de contador--: " << contador << "\n";

    ++contador; // incremento pre-fijo
    cout << "después de ++contador: " << contador << "\n";

    --contador; // decremento pre-fijo
    cout << "después de --contador: " << contador << "\n";

    // Diferencia pre-fijo vs post-fijo
    int x = 5;
    int y = ++x; // incrementa x primero, luego asigna (x es 6, y es 6)
    cout << "++x: x = " << x << ", y = " << y << "\n";

    int a = 5;
    int b = a++; // asigna a primero, luego incrementa (a es 6, b es 5)
    cout << "a++: a = " << a << ", b = " << b << "\n";

    // ===== OPERADOR TERNARIO =====

    cout << "\n===== OPERADOR TERNARIO =====\n";

    int edad = 18;
    string puedeVotar = (edad >= 18) ? "Sí, puede votar" : "No, aún es menor de edad";
    cout << "¿" << edad << " años puede votar? " << puedeVotar << "\n";

    int calificacion = 75;
    string resultado = (calificacion >= 60) ? "Aprobado" : "Reprobado";
    cout << "Calificación de " << calificacion << ": " << resultado << "\n";

    // ===== OPERADORES BITWISE =====

    cout << "\n===== OPERADORES BITWISE =====\n";

    int bitA = 5; // binario: 0101
    int bitB = 3; // binario: 0011

    int bitAnd = bitA & bitB; // AND bit a bit
    int bitOr = bitA | bitB; // OR bit a bit
    int bitXor = bitA ^ bitB; // XOR bit a bit
    int bitNot = ~bitA; // NOT bit a bit
    int leftShift = bitA << 1; // desplazamiento a la izquierda
    int rightShift = bitA >> 1; // desplazamiento a la derecha

    cout << bitA << " & " << bitB << " = " << bitAnd << "\n";
    cout << bitA << " | " << bitB << " = " << bitOr << "\n";
    cout << bitA << " ^ " << bitB << " = " << bitXor << "\n";
    cout << "~" << bitA << " = " << bitNot << "\n";
    cout << bitA << " << 1 = " << leftShift << "\n";
    cout << bitA << " >> 1 = " << rightShift << "\n";

    // ===== OPERADOR SIZEOF =====

    cout << "\n===== OPERADOR SIZEOF =====\n";

    int miEntero = 42;
    cout << "sizeof(miEntero): " << sizeof(miEntero) << " bytes\n";
    cout << "sizeof(int): " << sizeof(int) << " bytes\n";
    cout << "sizeof(double): " << sizeof(double) << " bytes\n";

    return 0;
}
